src/converters/downcast.js

```diff
@@ -0,0 +1,401 @@
+		conversionApi.writer.insert( viewPosition, asWidget ? tableWidget : tableElement );
+	}, { priority: 'normal' } );
+}
+
+/**
+ * Model row element to view <tr> element conversion helper.
+ *
+ * This conversion helper creates whole <tr> element with child elements.
+ *
+ * @returns {Function} Conversion helper.
+ */
+export function downcastInsertRow( options = {} ) {
+	return dispatcher => dispatcher.on( 'insert:tableRow', ( evt, data, conversionApi ) => {
+		const tableRow = data.item;
+
+		if ( !conversionApi.consumable.consume( tableRow, 'insert' ) ) {
+			return;
+		}
+
+		const table = tableRow.parent;
+
+		const tableElement = conversionApi.mapper.toViewElement( table );
+
+		const row = table.getChildIndex( tableRow );
+
+		const tableWalker = new TableWalker( table, { startRow: row, endRow: row } );
+
+		for ( const tableWalkerValue of tableWalker ) {
+			const tableSection = getOrCreateTableSection( getSectionName( tableWalkerValue ), tableElement, conversionApi );
+			const trElement = getOrCreateTr( tableRow, row, tableSection, conversionApi );
+
+			// Consume table cell - it will be always consumed as we convert whole row at once.
+			conversionApi.consumable.consume( tableWalkerValue.cell, 'insert' );
+
+			createViewTableCellElement( tableWalkerValue, ViewPosition.createAt( trElement, 'end' ), conversionApi, options );
+		}
+	}, { priority: 'normal' } );
+}
+
+/**
+ * Model row element to view <tr> element conversion helper.
+ *
+ * This conversion helper creates whole <tr> element with child elements.
+ *
+ * @returns {Function} Conversion helper.
+ */
+export function downcastInsertCell( options = {} ) {
+	return dispatcher => dispatcher.on( 'insert:tableCell', ( evt, data, conversionApi ) => {
+		const tableCell = data.item;
+
+		if ( !conversionApi.consumable.consume( tableCell, 'insert' ) ) {
+			return;
+		}
+
+		const tableRow = tableCell.parent;
+		const table = tableRow.parent;
+
+		const tableWalker = new TableWalker( table );
+
+		// We need to iterate over a table in order to get proper row & column values from a walker
+		for ( const tableWalkerValue of tableWalker ) {
+			if ( tableWalkerValue.cell === tableCell ) {
+				const trElement = conversionApi.mapper.toViewElement( tableRow );
+				const insertPosition = ViewPosition.createAt( trElement, tableRow.getChildIndex( tableCell ) );
+
+				createViewTableCellElement( tableWalkerValue, insertPosition, conversionApi, options );
+
+				// No need to iterate further.
+				return;
+			}
+		}
+	}, { priority: 'normal' } );
+}
+
+/**
+ * Conversion helper that acts on attribute change for headingColumns and headingRows attributes.
+ *
+ * Depending on changed attributes this converter will:
+ * - rename <td> to <th> elements or vice versa
+ * - create <thead> or <tbody> elements
+ * - remove empty <thead> or <tbody>
+ *
+ * @returns {Function} Conversion helper.
+ */
+export function downcastAttributeChange( options ) {
```

Comment:
yeah - splitting this to headingRows/ headingColumns probably will help. The complicated part is fixing table while chaning headingRows number. If a cell has rowspan and it could exceed `thead` it must be split beforehand as cells cannot go over thead.